Don't set ulimits (nproc) for all init scripts

thaJeztah opened the pull request:
There is a not-insignificant performance overhead for all containers (if
containerd is a child of Docker, which is the current setup) if rlimits are
set on the main Docker daemon process (because the limits
propogate to all children).

We recommend using cgroups to do container-local accounting.

This applies the change added in 8db61095a3d0bcb0733580734ba5d54bc27a614d
to other init scripts.

relates to https://github.com/docker/docker/pull/24307

ping @justincormack @tianon @runcom @cyphar PTAL if this looks good to you; I didn't know the best way to verify/test this, but can try if you agree on this

thaJeztah committed:
Don't set ulimits (nproc)

There is a not-insignificant performance overhead for all containers (if
containerd is a child of Docker, which is the current setup) if rlimits are
set on the main Docker daemon process (because the limits
propogate to all children).

We recommend using cgroups to do container-local accounting.

This applies the change added in 8db61095a3d0bcb0733580734ba5d54bc27a614d
to other init scripts.

Note that nfile cannot be set to unlimited, and the limit
is hardcoded to 1048576 (2^20) , see:
http://stackoverflow.com/a/1213069/1811501

Signed-off-by: Sebastiaan van Stijn <github@gone.nl>

tianon commented:
LGTM!

cyphar commented:
LGTM

cpuguy83 commented:
unlimited is still a number, just that it's a really high number.

thaJeztah commented:
@cpuguy83 is there a way to set no limit? If setting `unlimited` does still apply a limit, I wonder if https://github.com/docker/docker/pull/24307 (and this PR) has any benefits.

@cyphar did you benchmark the difference before/after applying https://github.com/docker/docker/pull/24307?

cyphar commented:
@thaJeztah According to the reporter of the bug, their benchmark time almost **halved** (from ~130 minutes to ~80) after setting the ulimits to be unlimited.

cyphar commented:
@cpuguy83 The special value `RLIM_INFINITY` is used by the kernel to mean "no limit". It is **not** the same as using a different very large number.

cyphar commented:
@cpuguy83 It looks like for `RLIMIT_NOFILE` Linux doesn't use `RLIM_INFINITY` as a special value specifically for the fastpath. So I'm not really sure _why_ there is a performance impact -- it might be some weird quirk of `rpm` that causes this to happen.

justincormack commented:
In fact you don't appear to be able to set `RLIMIT_NOFILE` to unlimited anyway:

```
ulimit -n unlimited
bash: ulimit: open files: cannot modify limit: Operation not permitted
```

It is bounded by `/proc/sys/fs/file-max`. So that line is not doing anything at all. So it can only be the process limit that is having an effect, if anything there will probably be a lower file limit than before as it was not being raised.

thaJeztah commented:
@justincormack hm, that's weird I think it tried that, but maybe I messed up and only tried on `ulimit -u` :blush: I see the [man page mentions](http://man7.org/linux/man-pages/man1/bash.1.html);

> -n     The maximum number of open file descriptors (most systems do not allow this value to be set)

Looks like you can only set a numeric value, that's *lower* or *equal* to 1048576, but not a higher one;

```
ulimit -n 1048576

ulimit -n 1048577
bash: ulimit: open files: cannot modify limit: Operation not permitted
```

but that limit doesn't match `/proc/sys/fs/file-max`;

```
cat /proc/sys/fs/file-max
398153
```

I'm trying that in a container, so possibly that's related there

thaJeztah commented:
So, does setting `LimitNOFILE=infinity` in systemd work? And why? Or is it silently ignored?

justincormack commented:
Try it outside a container, docker may already have that limit set...
On 13 Jul 2016 12:51 p.m., "Sebastiaan van Stijn" <notifications@github.com>
wrote:

> @justincormack <https://github.com/justincormack> hm, that's weird I
> think it tried that, but maybe I messed up and only tried on ulimit -u 😊
> I see the man page mentions
> <http://man7.org/linux/man-pages/man1/bash.1.html>;
>
> -n The maximum number of open file descriptors (most systems do not allow
> this value to be set)
>
> Looks like you can only set a numeric value, that's *lower* or *equal* to
> 1048576, but not a higher one;
>
> ulimit -n 1048576
>
> ulimit -n 1048577
> bash: ulimit: open files: cannot modify limit: Operation not permitted
>
> but that limit doesn't match /proc/sys/fs/file-max;
>
> cat /proc/sys/fs/file-max
> 398153
>
> I'm trying that in a container, so possibly that's related there
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/docker/docker/pull/24555#issuecomment-232332677>, or mute
> the thread
> <https://github.com/notifications/unsubscribe/AAdcPMGGjZNiOldl4tMntHJAMjmyyWjyks5qVNE-gaJpZM4JK1Uw>
> .
>

thaJeztah commented:
nope, looks to be the same;

```
root@dockr:~# ulimit -n 1048576
root@dockr:~# ulimit -n 1048577
-bash: ulimit: open files: cannot modify limit: Operation not permitted
root@dockr:~# ulimit -n unlimited
-bash: ulimit: open files: cannot modify limit: Operation not permitted
```

```
cat /proc/sys/fs/file-max
100809
```

justincormack commented:
ok that limit is hardcoded http://stackoverflow.com/questions/1212925/on-linux-set-maximum-open-files-to-unlimited-possible

thaJeztah commented:
Updated to only set `nproc`, and updated the systemd unit file accordingly; @cyphar can you confirm if setting `LimitNOFILE=infinity` actually worked? The systemd documentation seems to indicate it accepts the same values as `ulimit -n`; https://www.freedesktop.org/software/systemd/man/systemd.exec.html#id-1.7.2.29.17.3

justincormack commented:
LGTM

LK4D4 commented:
LGTM

cyphar commented:
@thaJeztah `LimitNOFILE=infinity` *definitely* works. I'm currently running a Docker daemon with that limit set.
